lib/src/config_document_parser.cc

```diff
@@ -0,0 +1,632 @@
+#include <internal/nodes/config_node_single_token.hpp>
+#include <internal/nodes/config_node_comment.hpp>
+#include <internal/nodes/config_node_concatenation.hpp>
+#include <internal/nodes/config_node_simple_value.hpp>
+#include <internal/path_parser.hpp>
+#include <internal/config_util.hpp>
+
+#include <unordered_map>
+#include <internal/nodes/config_node_field.hpp>
+#include <internal/nodes/config_node_array.hpp>
+
+using namespace std;
+
+namespace hocon { namespace config_document_parser {
+
+    shared_ptr<config_node_root> parse(token_iterator tokens, shared_origin origin,
+                                                                    config_parse_options options)
+    {
+        parse_context context { options.get_syntax(), move(origin), move(tokens) };
+        return context.parse();
+    }
+
+    shared_node_value parse_value(token_iterator tokens, shared_origin origin,
+                                                          config_parse_options options)
+    {
+        parse_context context { options.get_syntax(), move(origin), move(tokens) };
+        return context.parse_single_value();
+    }
+
+    /** Parse context */
+    parse_context::parse_context(config_syntax flavor, shared_origin origin, token_iterator tokens) :
+        _line_number(1), _tokens(move(tokens)), _flavor(flavor), _base_origin(move(origin)) { }
+
+    shared_token parse_context::pop_token() {
+        if (_buffer.empty()) {
+            return _tokens.next();
+        }
+        shared_token top = _buffer.top();
+        _buffer.pop();
+        return top;
+    }
+
+    shared_token parse_context::next_token() {
+        shared_token t = pop_token();
+        if (_flavor == config_syntax::JSON) {
+            if (t->get_token_type() == token_type::UNQUOTED_TEXT && !is_unquoted_whitespace(t)) {
+                throw parse_exception("Token not allowed in valid JSON: '" + t->token_text() + "'");
+            } else if (t->get_token_type() == token_type::SUBSTITUTION) {
+                throw parse_exception("Substitutions (${} syntax) not allowed in JSON");
+            }
+        }
+        return t;
+    }
+
+    shared_token parse_context::next_token_collecting_whitespace(shared_node_list& nodes) {
+        while (true) {
+            shared_token t = next_token();
+            if (t->get_token_type() == token_type::IGNORED_WHITESPACE || t->get_token_type() == token_type::NEWLINE
+                    || is_unquoted_whitespace(t)) {
+                nodes.push_back(make_shared<config_node_single_token>(t));
+                if (t->get_token_type() == token_type::NEWLINE) {
+                    _line_number = t->line_number() + 1;
+                }
+            } else if (t->get_token_type() == token_type::COMMENT) {
+                nodes.push_back(make_shared<config_node_comment>(t));
+            } else {
+                if (t->line_number() >= 0) {
+                    _line_number = t->line_number();
+                }
+                return t;
```

Comment:
I'm not sure that it does. Most of the no-more-tokens checking happens elsewhere; this class's loops tend to terminate on other things, and assume there will always be another token. If there isn't, usually it's because the stream of tokens is malformed, and an exception is thrown elsewhere.